src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -2548,6 +2647,326 @@ public final void addBatch() throws SQLServerException {
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 2);
+                return tempstr + "." + parseUserSQLForTableNameDW(true, true);
+            } else {
+                // return tablename
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 1);
+                return tempstr;
+            }
+        }
+        
+        // At this point, the next chunk of string is the table name (could have database name), without starting with [ or ".
+        StringBuilder sb = new StringBuilder();
+        while (localUserSQL.length() > 0) {
+            if (localUserSQL.charAt(0) == '.' || localUserSQL.charAt(0) == ' ' || localUserSQL.charAt(0) == '\t'
+                    || localUserSQL.charAt(0) == '(') {
+                if (localUserSQL.charAt(0) == '.') {
+                    localUserSQL = localUserSQL.substring(1);
+                    return sb.toString() + "." + parseUserSQLForTableNameDW(true, true);
+                } else {
+                    return sb.toString();
+                }
+            } else {
+                sb.append(localUserSQL.charAt(0));
+                localUserSQL = localUserSQL.substring(1);
+            }
+        }
+        
+        // It shouldn't come here. If we did, something is wrong.
+        throw new IllegalArgumentException("localUserSQL");
+    }
+    
+    private ArrayList<String> parseUserSQLForColumnListDW() {
+        localUserSQL = localUserSQL.trim();
+        
+        // ignore all comments
+        if (checkAndRemoveComments()) {
+            return parseUserSQLForColumnListDW();
+        }
+        
+        //check if optional column list was provided
+        // Columns can have the form of c1, [c1] or "c1". It can escape ] or " by ]] or "".
+        if (localUserSQL.substring(0, 1).equalsIgnoreCase("(")) {
+            localUserSQL = localUserSQL.substring(1);
+            return parseUserSQLForColumnListDWHelper(new ArrayList<String>());
+        }
+        return null;
+    }
+
+    private ArrayList<String> parseUserSQLForColumnListDWHelper(ArrayList<String> listOfColumns) {
+        localUserSQL = localUserSQL.trim();
+        
+        // ignore all comments
+        if (checkAndRemoveComments()) {
+            return parseUserSQLForColumnListDWHelper(listOfColumns);
+        }
+        
+        if (localUserSQL.charAt(0) == ')') {
+            localUserSQL = localUserSQL.substring(1);
+            return listOfColumns;
+        }
+        
+        if (localUserSQL.charAt(0) == ',') {
+            localUserSQL = localUserSQL.substring(1);
+            return parseUserSQLForColumnListDWHelper(listOfColumns);
+        }
+        
+        if (localUserSQL.charAt(0) == '[') {
+            int tempint = localUserSQL.indexOf("]", 1);
+            
+            // keep checking if it's escaped
```

Comment:
I guess the parsing would fail if we have `[aaaa]]]]bbbbb]` as a column name too?

Reply:
the column name parsing is done correctly, but I had to make some changes for it down the line (which also applies to bulk copy from CSV).